packages/toolpad-app/src/toolpad/AppEditor/BindingEditor.tsx

```diff
@@ -501,18 +504,22 @@ export function BindingEditorDialog<V>({
 
   const committedInput = React.useRef<BindableAttrValue<V> | null>(input);
 
-  const handleSave = React.useCallback(() => {
+  const handleSave = React.useCallback(async () => {
     let newValue = input;
 
     if ((input as JsExpressionAttrValue)?.$$jsExpression) {
+      const jsExpression = await tryFormatExpression(
+        (input as JsExpressionAttrValue).$$jsExpression,
+        data!,
```

Comment:
If the backend call to `getPrettierConfig` fails, or is still loading, `data` will be `undefined`. So using `!` is not the right solution as that is only reserved for when `data` is guaranteed to be defined by the programmer.

Instead, show a warning to the user that explains that the prettier config couldn't be loaded and therefore the code won't be formatted when saving.

Reply:
> If the backend call to `getPrettierConfig` fails, or is still loading, `data` will be `undefined`. So using `!` is not the right solution as that is only reserved for when `data` is guaranteed to be defined by the programmer.
> 
> Instead, show a warning to the user that explains that the prettier config couldn't be loaded and therefore the code won't be formatted when saving.

what's type of warning for user?
such as https://mui.com/material-ui/react-alert/?

Reply:
is there component that i can directly invoke ?

Reply:
An Alert could work. I'd have to see what it looks like. For me I think it's important it's inline, close enough to the save button (i.e. no popup)

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/64976b65-ab33-4804-9c86-c8c682256aa1)
I checked again.
user want to reformat the code but that code doesn't be formated because the network signal not well.
and then user must reload the website.
It will bring a new problem. such as: If the user wishes to format code without reloading the page.

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/cbb558d7-5152-4dee-a4fb-67b482994329)
how about this ?

Reply:
>will do as to give more clues about the reason?

I'm not totally understand. what's clue? 

this error only happens without a network I think

Reply:
![Uploading image.png…]()

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/2486d23f-a2ba-43bd-a860-37e892ca019e)

Reply:
> this error only happens without a network I think

There could be a problem while prettier tries to access the file system, or a bug in our code, or in prettier, or our bundler could have transformed the code wrong, or it could be running on an incompatible node.js runtime that we're unaware about, a solar storm flipping a bit,... anything really 🙂

Reply:
![image](https://github.com/mui/mui-toolpad/assets/27261522/bbc993c9-babc-44ae-a84d-09e59fb7be1e)
```js
const { data } = client.useQuery('getPrettierConfig', []);
```
I see.

do these errors need to be passed to front-end, if back-end occurs error ?

and also I didn't see any `try catch` in reading prettier config

Reply:
@Janpot
please make a final check.

Reply:
@JerryWu1234 [`useQuery`](https://github.com/mui/mui-toolpad/pull/2674/files#diff-67eb614d84fd465e47f1730731e8e1cef4e2e15dcf3294bcfc35546866a82b2cR508) is a wrapper around `react-query`. Next to the `data` property it will return an `error` property.